core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala

```diff
@@ -550,22 +550,26 @@ import com.github.ghik.silencer.silent
         }
       }
 
-      // The strategy is simple, if full partition was cleaned up we will find null-row
-      // i.e row: <persistence_id>, <partition_nr>, null, ...
-      // If such row exists, we have to switch partition and continue to fetching
-      def checkForGaps(): Unit = {
+      // See PR #509 for background
+      // Only used when gapFreeSequenceNumbers==false
+      // if full partition was cleaned up we look for two empty partitions before completing
+      def checkForGaps(foundEmptyPartitionCount: Int): Unit = {
```

Comment:
This was causing me most trouble, see #509
I'm still not 100% that this is correct in all cases, but it passes the test and the feature is not enabled by default.
I think the feature of removing arbitrary events without updating the `deleted_to` is questionable.

Reply:
Agreed